library/core/src/iter/adapters/chain.rs

```diff
@@ -290,3 +291,41 @@ fn and_then_or_clear<T, U>(opt: &mut Option<T>, f: impl FnOnce(&mut T) -> Option
     }
     x
 }
+
+#[rustc_unsafe_specialization_marker]
+trait SymmetricalArms {}
+
+impl<A> SymmetricalArms for Chain<A, A> {}
+
+trait SpecChain: Iterator {
+    fn next(&mut self) -> Option<Self::Item>;
+}
+
+impl<A, B> SpecChain for Chain<A, B>
+where
+    A: Iterator,
+    B: Iterator<Item = A::Item>,
+{
+    #[inline]
+    default fn next(&mut self) -> Option<A::Item> {
+        and_then_or_clear(&mut self.a, Iterator::next).or_else(|| self.b.as_mut()?.next())
+    }
+}
+
+impl<A, B> SpecChain for Chain<A, B>
+where
+    A: Iterator,
+    B: Iterator<Item = A::Item>,
+    Self: SymmetricalArms,
+{
+    #[inline]
+    fn next(&mut self) -> Option<A::Item> {
+        let mut result = and_then_or_clear(&mut self.a, Iterator::next);
+        if result.is_none() {
+            // SAFETY: SymmetricalArms guarantees that A and B are the same type.
```

Comment:
This is not true. Specialization is not able to determine whether two arbitrary types are equal to each other if lifetimes are involved. If you have `Chain<Foo<'static>, Foo<'a>>` this code is going to transmute `Foo<'a>` to `Foo<'static>`. 

Since `Chain` hands out references to the fields to safe code via `Iterator` `Clone` and `Debug` methods this is almost certainly an unsound optimization (which would be observable by safe code in ways that are only possible via specialization ANYWAY even if this was sound :woman_shrugging:)

Reply:
But if `Chain` is composed of those Foos then its own lifetime must be shorter than both, and thus any borrow of its fields.
How would `Foo` weaponize the difference between `'static` and `'a`?

Reply:
Well, Copy is `#[rustc_unsafe_specialization_marker]` too which means you're not allowed to implement it in a lifetime-dependent way. And that's already an issue on its own, it doesn't need Chain: https://github.com/rust-lang/rust/issues/88901#issuecomment-918336437
SymmetricalArms isn't public and so can't be implemented in a lifetime-dependent way.

So that's more pointing out that Copy has a soundness hole rather than SymmetricalArms having one.

My understanding is that if we have both `Foo<'a>` and `Foo<'static>` and both fulfill the same trait bounds and we only use those traits then monomoriphization will emit the same code for both because all the trait implementations we're using are lifetime-independent because they're shared between the two.

Reply:
> So that's more pointing out that Copy has a soundness hole rather than SymmetricalArms having one.

I'm not sure that distinction is useful. `Copy` is a way to "weaponize" the difference, but it's still incorrect treat `Chain<Foo<'static>, Foo<'a>>` (or any other distinct lifetimes) as `SymmetricalArms`.

Reply:
> but it's still incorrect treat Chain<Foo<'static>, Foo<'a>> (or any other distinct lifetimes) as SymmetricalArms.

I mean we can call it `SymmetricalModuloLifetimes` if you want. My point is that lifetimes shouldn't matter.

* lifetimes don't matter for evaluating types because `Chain<A, B>` places the same bounds on A and B which means all of the used traits are not conditional on lifetimes
* lifetimes are erased at codegen time
* punning is valid at runtime (the compiler makes use of that!) so long one ensures that the values don't exceed their actual lifetime. This is ensured by Chain's combined lifetime being the shorter of the two.

Yes, having a general mechanism to turn a `Foo<'a>` into `Foo<'static>` would be wrong. But I think this particular instance it's contained safely.

Reply:
> There are stable observable differences due to std's specialization, e.g. if you implement `Clone` for all lifetimes and `Copy` only for `'static`, then methods like `slice::fill` behave differently.

I misunderstood this -- I thought that such `Copy` specialization would let you observe the difference before and after this PR. But no, that's already broken -- the existence of `impl Copy for Foo<'static>` already makes those specializations apply for all `Foo<'_>`.

Reply:
I've marked the trait as unsafe and added comments explaining why it's currently ok to do this.